libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,500 @@
+//! # Buildpack output
+//!
+//! Use the [`BuildpackOutput`] to output structured text as a buildpack is executing.
+//!
+//! ```
+//! use libherokubuildpack::buildpack_output::BuildpackOutput;
+//!
+//! let mut output = BuildpackOutput::new(std::io::stdout())
+//!     .start("Heroku Ruby Buildpack")
+//!     .warning("No Gemfile.lock found");
+//!
+//! output = output
+//!     .section("Ruby version")
+//!     .finish();
+//!
+//! output.finish();
+//! ```
+//!
+use crate::buildpack_output::ansi_escape::{BOLD_PURPLE, CYAN, RED, YELLOW};
+use crate::buildpack_output::util::{prefix_first_rest_lines, prefix_lines, ParagraphInspectWrite};
+use crate::write::line_mapped;
+use std::fmt::Debug;
+use std::io::Write;
+use std::time::Instant;
+
+mod ansi_escape;
+mod duration_format;
+pub mod style;
+mod util;
+
+/// See the module docs for example usage.
```

Comment:
For UX on docs.rs and in an IDE, what do you think about duplicating the module docs here? (Longer term I think it's fine for both sections to differ - eg one of the two to be the canonical location for full examples, and the other a short example.)

Otherwise I get this in my IDE:

<img width="327" alt="Screenshot 2024-02-02 at 15 32 07" src="https://github.com/heroku/libcnb.rs/assets/501702/1ccaf545-bf00-4c7e-9a38-8aad8aa47275">

And this in the generated rustdocs:

<img width="1209" alt="Screenshot 2024-02-02 at 15 33 58" src="https://github.com/heroku/libcnb.rs/assets/501702/5f173d5b-d9f1-468c-b392-f62e37a545a2">

Reply:
I think we could move it to an external document, like a README.md in the folder than use str include like https://github.com/schneems/magic_migrate/blob/3e43a72c329a2f03c016e0d37cd8ec085af7982f/src/lib.rs#L1 (though I'm not 100% sure that also works on a struct doc, but I'll try).

Reply:
I've updated the docs, please take a look.

Reply:
I'm not sure I like splitting out the `build_output` module docs to a `README.md`. The only reason we do that for the crate docs, is that we have to have a README.md for crates.io / GitHub, so it makes sense to combine the two (and we want to have the README examples covered in CI via doctests). 

For the `build_output` case, it means a README nested in a random directory, and means people reading the Rust code have to switch to another file to see the contents.

Also, I think longer term it's likely we'll want different contents for the module vs struct docs, so there won't be the need to de-duplicate the two. Plus duplication isn't really a problem anyway IMO.

Reply:
I put the readme in the module docs and duplicated the example to the BuildOutput struct